### DISCUSSION TITLE
How to enable code spelling check?

### DISCUSSION THREAD

**Discussion by alexsmartens** [tree-sitter/tree-sitter#3444] [Q&A] [answered]
Hi all,
I am looking for a way to enable code spell checking, specifically variable, class and method names. Would anyone have suggestions how to configure code spelling check?

For reference, a basic Neovim configuration only enables spell check for comments, like this:
```
local opt = vim.opt

opt.spell = true
opt.spelllang = { "en_us" }
opt.spelloptions:append({ "camel" })
```

In https://github.com/nvim-treesitter/nvim-treesitter/issues/5202#issuecomment-1667434770, it seems like @ShellCode33 has enabled code spell checking.

Based on these discussions, it seems like there is a way to enable `@spell` for code nodes which would trigger code spelling:
 * @lucario387, @sonic182: https://github.com/nvim-treesitter/nvim-treesitter/issues/6057#issuecomment-1931880635
 * @Hacksore, @clason https://github.com/nvim-treesitter/nvim-treesitter/issues/6082
 * https://github.com/nvim-treesitter/nvim-treesitter/blob/7ff0b8d3d21b15eb7231fcbd2acfded8987f5efd/CONTRIBUTING.md?plain=1#L260

I found a few references pointing to [spellsitter.nvim](https://github.com/lewis6991/spellsitter.nvim) which is merged into Neovim core at this point. However, I also cannot find any configuration example to fit my bill.
 
 @lewis6991, @vigoux I noticed you commented on related issues, perhaps you would know how to enable code spelling check.

**Comment by alexsmartens** [comment, marked as answer]
I think I got it to work, more details in this commit https://github.com/alexsmartens/dotfiles/commit/9dd2f8da2f9e112077267e0c84b48248f076484f.

In the nutshell:
* enable neovim spelling check:
```lua
-- Enable syntax highlighting
vim.cmd("syntax on")
-- Enable spellcheck
opt.spell = true
opt.spelllang = { "en_us" }
opt.spelloptions:append "camel"
opt.spellcapcheck = "" -- disable checking for capital letters at the start of sentences
```
* enable `nvim-treesitter` [spell checking](https://github.com/nvim-treesitter/nvim-treesitter/blob/7ff0b8d3d21b15eb7231fcbd2acfded8987f5efd/CONTRIBUTING.md?plain=1#L260) for all nodes: 
```scm
; Spell check each node
 ((identifier) @spell)
 ; Spell check other nodes that are not being checked for some reason
 ((constant) @spell)
 ((string) @spell)

; Copy the default config's content below or otherwise all the default configurations would be lost
; https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/ruby/highlights.scm
```

Done!

**Reply by sogaiu** [reply]
Is it correct that this is something specific to nvim-treesitter?

If so, it might be more useful for it to live over that repository.

**Reply by alexsmartens** [reply]
oops. Yup, it should leave there

**Reply by alexsmartens** [reply]
For reference, I opened another discussion https://github.com/nvim-treesitter/nvim-treesitter/discussions/6877

**Comment by alexsmartens** [comment]
To verify if an identifier (e.g., a variable name) is being spell checked:
* hover the cursor over your identifier;
* run `:Inspect` to inspect the syntax tree and highlight the node under the cursor.

In the screenshot below, `spell` on line `7` is the node I am looking at. And `@spell` is enabled for it.

![Screenshot 2024-07-05 at 4 29 14 PM](https://github.com/tree-sitter/tree-sitter/assets/26261714/fc0c358b-1dac-4167-8fd0-61bd448c60a8)
